feat: [UIE-8135] - IAM RBAC: add new assign panel component

aaleksee-akamai opened the pull request:
## Description 📝

IAM RBAC - new assign panel component


## Changes  🔄

List any change(s) relevant to the reviewer.

- new assign panel component
- fixing styles for `Permissions` and `Entities` components
- fixing a bug in the `Permissions` component when there are fewer than two lines of items

## Target release date 🗓️

2/11/25 (dev)


## Preview 📷

**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| ![image](https://github.com/user-attachments/assets/9363e9cf-3cbc-4fb2-b402-7a61c3f389ae) | ![image](https://github.com/user-attachments/assets/f0f27383-df00-4611-9511-ba3297051f85) |
| ![image](https://github.com/user-attachments/assets/b343dd5a-513a-4a61-aec3-4e42c5e44e5c) | ![image](https://github.com/user-attachments/assets/31b49356-914a-40e3-8f8a-3982c3c8a116) |


## How to test 🧪

### Prerequisites

(How to setup test environment)

- Ensure the Identity and Access Beta flag is enabled in dev tools
- Ensure the MSW is enabled in dev tools
- Click on the any username in the users table and go to the tab Assigned Roles or click on the user's menu View Assigned Roles
- Click on the button `Assign New Roles`
- Select any role from the dropdown

### Verification steps

(How to verify changes)

- [x]  Confirm component renders


<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

aaleksee-akamai commented on packages/manager/src/factories/accountPermissions.ts:
```diff
@@ -26,32 +26,42 @@ const createResourceRoles = (
+          permissions: viewer as PermissionType[],
+        }
+      : null,
+  ].filter(Boolean),
```
Remove null/undefined values

aaleksee-akamai commented on packages/manager/src/features/IAM/Shared/Permissions/Permissions.tsx:
```diff
@@ -49,9 +49,8 @@ export const Permissions = ({ permissions }: Props) => {
-      ? itemsArray.length - firstHiddenIndex
-      : 0;
+    const numHiddenItems =
+      firstHiddenIndex !== -1 ? itemsArray.length - firstHiddenIndex : 0;
```
this is a fix when the permissions are fewer than two lines and we have `firstHiddenIndex` equals -1 (which will be true then)

abailly-akamai commented on packages/manager/src/factories/accountPermissions.ts:
```diff
@@ -26,32 +26,42 @@ const createResourceRoles = (
+      ? {
+          description: `Access to perform any supported action on all ${resourceType} instances`,
+          name: `account_${resourceType}_admin`,
+          permissions: accountAdmin as PermissionType[],
```
I know that code existed before, but we don't really need to cast `as PermissionType[]` everywhere if doing

```TS
interface CreateResourceRoles {
  accountAdmin?: PermissionType[];
  [...]
}
```

aaleksee-akamai replied:
thanks, removed it

abailly-akamai commented on packages/manager/src/features/IAM/Shared/AssignedPermissionsPanel/AssignedPermissionsPanel.tsx:
```diff
@@ -0,0 +1,60 @@
+          theme.name === 'light'
+            ? theme.tokens.color.Neutrals[5]
+            : theme.tokens.color.Neutrals[90],
+        marginTop: 1.25,
```
optional: use `theme.spacing` when possible to make spacing consistent through our features

abailly-akamai commented on packages/manager/src/features/IAM/Users/UserRoles/AssignNewRoleDrawer.tsx:
```diff
@@ -0,0 +1,124 @@
+        Select a role you want to assign to a user. Some roles require selecting
+        resources they should apply to. Configure the first role and continue
+        adding roles or save the assignment.
+        <Link to=""> Learn more about roles and permissions.</Link>
```
Optional: Add a `// TODO` to remember to add a link in there (easy to forget as the feature grows)

abailly-akamai commented on packages/manager/src/features/IAM/Users/UserRoles/AssignNewRoleDrawer.tsx:
```diff
@@ -0,0 +1,124 @@
+    }
+  }
+  return null;
+};
```
Can we put those two functions in a `utils.ts` file and add coverage?

aaleksee-akamai replied:
sure, done

abailly-akamai reviewed (commented):
Thanks for the contribution! PR looks great overall, added some code improvement suggestions.

Additionally, I noticed a few things in the UI that could use your attention (some of those may be by design, some we may want to check with UX)

![Screenshot 2025-02-05 at 09 07 40](https://github.com/user-attachments/assets/57a3538b-95ac-4e8f-8811-2ce1e1a55bb9)

![Screenshot 2025-02-05 at 09 41 53](https://github.com/user-attachments/assets/d24f78e1-60f6-4956-bc91-4c915c610137)

![Screenshot 2025-02-05 at 09 31 00](https://github.com/user-attachments/assets/0e7ebe31-9bcb-4216-b03e-ad4106be09a0)

![Screenshot 2025-02-05 at 09 41 35](https://github.com/user-attachments/assets/775acc36-cfa6-4f61-892b-00fc3b9fd19b)

aaleksee-akamai commented:
> Thanks for the contribution! PR looks great overall, added some code improvement suggestions.
> 
> Additionally, I noticed a few things in the UI that could use your attention (some of those may be by design, some we may want to check with UX)
> 
> ![Screenshot 2025-02-05 at 09 07 40](https://private-user-images.githubusercontent.com/130582365/410042727-57a3538b-95ac-4e8f-8811-2ce1e1a55bb9.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDc4OTEsIm5iZiI6MTczODg0NzU5MSwicGF0aCI6Ii8xMzA1ODIzNjUvNDEwMDQyNzI3LTU3YTM1MzhiLTk1YWMtNGU4Zi04ODExLTJjZTFlMWE1NWJiOS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQxMzEzMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMWJiZGM2YTIyYWYxZTFhZmFjZjJhNzM0ZmE5OGVhMzJlODZmY2Q5ODUwN2FlMDU0ZTNlODYwZTE2MDFiMzAwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._se6IP36KlsP-Xjeze3ZoJ8mDtAhNp-zLCtzYkBle2M)
> 
> ![Screenshot 2025-02-05 at 09 41 53](https://private-user-images.githubusercontent.com/130582365/410042982-d24f78e1-60f6-4956-bc91-4c915c610137.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDc4OTEsIm5iZiI6MTczODg0NzU5MSwicGF0aCI6Ii8xMzA1ODIzNjUvNDEwMDQyOTgyLWQyNGY3OGUxLTYwZjYtNDk1Ni1iYzkxLTRjOTE1YzYxMDEzNy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQxMzEzMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNTA3NTNhN2FjNzViNWEyZGZmODMwMzVhODg1NTRiNzc1YjdlZDczZDBkMmY3ZTYyY2E0Yjk3ZjdmZWRhZmQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cNRVAPD8ZJYimXuxFTLw3vwXr_OAVtmiQldYIiFoy0I)
> 
> ![Screenshot 2025-02-05 at 09 31 00](https://private-user-images.githubusercontent.com/130582365/410043681-0e7ebe31-9bcb-4216-b03e-ad4106be09a0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDc4OTEsIm5iZiI6MTczODg0NzU5MSwicGF0aCI6Ii8xMzA1ODIzNjUvNDEwMDQzNjgxLTBlN2ViZTMxLTliY2ItNDIxNi1iMDNlLWFkNDEwNmJlMDlhMC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQxMzEzMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZGVhOGU2N2Y3ODNiNTdiMWUyMWI3ODQ4ZTAwMmIyYmE2MWYyZjViYmFiMjk2YWEwNGY0NGMxOTg0NTVlNTE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.V67gmrcYAFQ3e-5sI6JpKmbjRFjmeLCvhE3qAcPAaxw)
> 
> ![Screenshot 2025-02-05 at 09 41 35](https://private-user-images.githubusercontent.com/130582365/410043953-775acc36-cfa6-4f61-892b-00fc3b9fd19b.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4NDc4OTEsIm5iZiI6MTczODg0NzU5MSwicGF0aCI6Ii8xMzA1ODIzNjUvNDEwMDQzOTUzLTc3NWFjYzM2LWNmYTYtNGY2MS04OTJiLTAwZmMzYjlmZDE5Yi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwNlQxMzEzMTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZGYwMGI3NmYxY2YwYzQ4NWQwNWIwNThhNTMxMjVkOTg4ZmVhZTMwMThlOGM2MTlkYWZmMGQyYzAxNjkxYjUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UWpbQ8k_1bb24c4V3jC0Bv62JrU8fK5lJDZn9GblXGo)

Thanks @abailly-akamai ! I've fixed the issue with the autocomplete and colors, about the right border - we'll discuss that with the designers

abailly-akamai reviewed (approved):
Thanks for addressing the feedback - confirmed the Autocomplete and overall functionality is looking good as far as my testing went ✅

cpathipa reviewed (approved):
Nice work @aaleksee-akamai. The code and functionality LGTM!
